src/clojure/nrepl/cmdline.clj

```diff
@@ -239,59 +244,193 @@
+  "Prints the nREPL server header that some tools rely on to parse the
+  connection details from.
+  Takes nREPL server map and processed CLI options map.
+  Returns nil."
+  [server options]
+  (let [transport (:transport options)
+        port (:port server)
+        ^java.net.ServerSocket ssocket (:server-socket server)
+        host (.getHostName (.getInetAddress ssocket))]
+    ;; The format here is important, as some tools (e.g. CIDER) parse the string
+    ;; to extract from it the host and the port to connect to
+    (println (format "nREPL server started on port %d on host %s - %s://%s:%d"
+                     port host (transport/uri-scheme transport) host port))))
+
+(defn save-port-file!
+  "Writes a file relative to project classpath with port number so other tools
+  can infer the nREPL server port.
+  Takes nREPL server map and processed CLI options map.
+  Returns nil."
+  [server options]
+  ;; Many clients look for this file to infer the port to connect to
+  (let [port (:port server)
+        port-file (io/file ".nrepl-port")]
+    (.deleteOnExit port-file)
+    (spit port-file port)))
+
+(defn interactive-repl-or-sleep
+  "Runs an interactive repl if :interactive CLI option is true otherwise
+  puts the current thread to sleep
+  Takes nREPL server map and processed CLI options map.
+  Returns nil."
+  [server options]
+  (if (:interactive options)
+    (let [transport (:transport options)
+          host (:host server)
+          port (:port server)]
+      (run-repl host port (merge (when (:color options) colored-output)
+                                 {:transport transport})))
+      ;; need to hold process open with a non-daemon thread
+      ;;   -- this should end up being super-temporary
+    (Thread/sleep Long/MAX_VALUE)))
+
+(defn server-command
+  "Creates an nREPL server instance, prints connection info, saves port file,
+  and runs an interactive repl or puts the thread to sleep.
+  Takes map of CLI options.
+  Returns nREPL server map."
+  [options]
+  (let [options (get-server-options options)
+        {:keys [port bind handler transport greeting]} options
+        server (start-server :port port
+                             :bind bind
+                             :handler handler
+                             :transport-fn transport
+                             :greeting-fn greeting)]
+    (doto server
+      (ack-server options)
+      (print-connection-header options)
+      (save-port-file! options)
+      (interactive-repl-or-sleep options))))
+
+(defn dispatch-commands
+  "Look at options to dispatch a specified command.
+  Takes CLI options map. May return a server map, nil, or exit."
+  [options]
+  (cond (:help options)    (help-command)
+        (:version options) (version-command)
+        (:connect options) (connect-command options)
+        :else (server-command options)))
+
 (defn- run
+  "Parses args, dispatches commands like help, version, connect, or starts an
```

Comment:
I wonder if we need this `run` function at all. I guess it can simply go into `main` now.